core/src/main/java/io/smallrye/openapi/api/models/MapBasedModelImpl.java

```diff
@@ -0,0 +1,167 @@
+                } else if (oldValue instanceof ModelImpl) {
+                    into.put(name, MergeUtil.mergeObjects(oldValue, value));
+                } else {
+                    into.put(name, value);
+                }
+            }
+        }
+    }
+
+    private static <T> void mergeList(List<T> oldValue, List<T> value) {
+        Set<T> contents = new HashSet<>(oldValue);
+        for (T element : value) {
+            if (!contents.contains(element)) {
+                oldValue.add(element);
+            }
+        }
+    }
+
+    protected <T> void setProperty(String propertyName, T value) {
+        if (value == null) {
+            data.remove(propertyName);
+        } else {
+            data.put(propertyName, value);
+        }
+    }
+
+    protected <T> T getProperty(String propertyName, Class<T> type) {
+        Object result = data.get(propertyName);
+        if (type.isInstance(result)) {
+            return type.cast(result);
+        } else {
+            return null;
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    protected <T> List<T> getListProperty(String propertyName) {
+        Object result = data.get(propertyName);
+        if (result instanceof List) {
+            return Collections.unmodifiableList((List<T>) result);
+        } else {
+            return null;
+        }
+    }
+
+    protected <T> void setListProperty(String propertyName, List<T> value) {
+        value = ModelUtil.replace(value, ArrayList::new);
+        if (value == null) {
+            data.remove(propertyName);
+        } else {
+            data.put(propertyName, value);
+        }
+    }
+
+    protected <T> void addToListProperty(String propertyName, T value) {
+        List<T> list = (List<T>) data.get(propertyName);
+        if (value != null) {
+            if (list == null) {
+                list = new ArrayList<>();
+                data.put(propertyName, list);
+            }
+            list.add(value);
+        }
+    }
+
+    protected <T> void removeFromListProperty(String propertyName, T toRemove) {
+        List<T> list = (List<T>) data.get(propertyName);
+        ModelUtil.remove(list, toRemove);
+    }
+
+    protected <T> void setMapProperty(String propertyName, Map<String, T> value) {
+        value = ModelUtil.replace(value, HashMap::new);
+        if (value == null) {
+            data.remove(propertyName);
+        } else {
+            data.put(propertyName, value);
+        }
+    }
+
+    protected <T> Map<String, T> getMapProperty(String propertyName) {
+        Object result = data.get(propertyName);
+        if (result instanceof Map) {
+            return Collections.unmodifiableMap((Map<String, T>) data.get(propertyName));
+        } else {
+            return null;
+        }
+    }
+
+    protected <T> void addToMapProperty(String propertyName, String key, T value) {
+        if (value != null) {
+            Map<String, T> map = (Map<String, T>) data.get(propertyName);
```

Comment:
The add and remove methods for maps and lists risk a `ClassCastException` if the data stored in the property is not actually a map or a list.

We may want to improve this by:
* providing a better exception
* avoiding an exception
  * for add: overwriting the value with a new list or map
  * for remove: doing nothing

Reply:
I've implemented the second option.